src/api/wallet/WalletApiImpl.ts

```diff
@@ -0,0 +1,254 @@
+    log('[WalletApi] Connected')
+
+    await this._notifyListeners()
+
+    const subscriptions = createSubscriptions(provider)
+
+    const unsubscribeUpdates = subscribeToBlockchainUpdate({ subscriptions, provider, web3: this._web3 })(
+      this._notifyListeners.bind(this),
+    )
+
+    let unsubscribeDisconnect: Command = () => {}
+    if (isWalletConnectSubscriptions(subscriptions)) {
+      unsubscribeDisconnect = subscriptions.onStop(this.disconnect.bind(this))
+    } else if (isMetamaskSubscriptions(subscriptions)) {
+      unsubscribeDisconnect = subscriptions.onAccountsChanged(accounts => {
+        if (accounts.length > 0) return
+        // accounts  = [] when user locks Metamask
+
+        this.disconnect()
+      })
+    }
+
+    this._unsubscribe = (): void => {
+      unsubscribeUpdates()
+      unsubscribeDisconnect()
+    }
+
+    return true
+  }
+
+  public async disconnect(): Promise<void> {
+    if (isWalletConnectProvider(this._provider) && this._connected) await this._provider.close()
+
+    this._unsubscribe()
+
+    this._provider = null
+    this._web3 = null
+
+    log('[WalletApi] Disconnected')
+    await this._notifyListeners()
+  }
+
+  public async getAddress(): Promise<string> {
+    assert(this._connected, 'The wallet is not connected')
+
+    return this._user
+  }
+
+  public async getBalance(): Promise<BN> {
+    assert(this._connected, 'The wallet is not connected')
+
+    return toBN(await this._balance)
+  }
+
+  public async getNetworkId(): Promise<number> {
+    assert(this._connected, 'The wallet is not connected')
+
+    return this._networkId
+  }
+
+  public addOnChangeWalletInfo(callback: OnChangeWalletInfo, trigger?: boolean): Command {
+    this._listeners.push(callback)
+    if (trigger) {
+      callback(this._getWalletInfo())
+    }
+
+    return (): void => this.removeOnChangeWalletInfo(callback)
+  }
+
+  public removeOnChangeWalletInfo(callback: OnChangeWalletInfo): void {
+    this._listeners = this._listeners.filter(c => c !== callback)
+  }
+
+  /* ****************      Private Functions      **************** */
+
+  private _getWalletInfo(): WalletInfo {
+    const { isConnected = false, accounts = [], chainId = 0 } = getProviderState(this._provider) || {}
+    return {
+      isConnected,
+      userAddress: accounts[0],
+      networkId: isConnected ? +chainId : undefined,
+    }
+  }
+
+  private async _notifyListeners(): Promise<void> {
+    await Promise.resolve()
+    const walletInfo: WalletInfo = this._getWalletInfo()
+    console.table(walletInfo)
+    this._listeners.forEach(listener => listener(walletInfo))
+  }
+
+  private get _connected(): boolean {
+    return !!(getProviderState(this._provider) || {}).isConnected
+  }
+  private get _user(): string {
+    const { accounts: [account] = [] } = getProviderState(this._provider) || {}
+    return account
+  }
+  private get _balance(): Promise<string> {
+    if (!this._web3) return Promise.resolve('0')
```

Comment:
Why the balance is zero? if there's not connection I understand it should be null or undefined?

Reply:
Because it's specified in `WalletApi`. So someone somewhere decided we should have a defined value always.
If we want to change that now should be a separate issue.

Reply:
Ok, let's leave it for now then